### PR TITLE
Fixes #26728 - configure default language (and timezone)

### PR DESCRIPTION
**Pull request by sbernhard**

### CODE REVIEW

File: app/models/setting/general.rb [hunk 1]
```diff
@@ -22,7 +22,8 @@ def self.default_settings
       self.set('http_proxy_except_list', N_('Set hostnames to which requests are not to be proxied. Requests to the local host are excluded by default.'), [], N_('HTTP(S) proxy except hosts')),
       self.set('lab_features', N_("Whether or not to show a menu to access experimental lab features (requires reload of page)"), false, N_('Show Experimental Labs')),
       self.set("append_domain_name_for_hosts", N_("Foreman will append domain names when new hosts are provisioned"), true, N_("Append domain names to the host")),
-      self.set('outofsync_interval', N_("Duration in minutes after servers are classed as out of sync."), 30, N_('Out of sync interval'))
+      self.set('outofsync_interval', N_("Duration in minutes after servers are classed as out of sync."), 30, N_('Out of sync interval')),
+      self.set('default_language', N_("Language to use for new users"), 'Browser locale', N_('Default language'), nil, { :collection => Proc.new { Hash[FastGettext.human_available_locales.map { |lang| [lang[1], lang[0]] }].merge({ 'Browser locale' => 'Browser locale' }) } })
```
Comment (timogoebel) [inline app/models/setting/general.rb]: I guess you'll want to translate the string here. Also you don't need `{}` when calling merge.

```ruby
.merge('Browser locale' => _('Browser locale'))
```

**Comment by sbernhard** [comment]
Some questions:
1. is this something, foreman should have?
2. should I extend it to have default for timezone, too?
3. how to fix the katello issues? is it somehow possible to specify in this PR, to use a specific katello PR for running the tests?

File: app/models/user.rb [hunk 2]
```diff
@@ -689,6 +690,10 @@ def ensure_default_role
     self.roles << default_role unless self.roles.include?(default_role)
   end
 
+  def set_default_language_for_new_user
+    self.locale = Setting[:default_language] if self.respond_to?(:locale) && Setting[:default_language].present?
```
Comment (ezr-ondrej) [inline app/models/user.rb]: Is it necessary for some reason to check if we `respond_to?(:locale)`?
```suggestion
    self.locale ||= Setting[:default_language].presence
```
The only important part here for me is not to overwrite the value if there is one.

**Review by ezr-ondrej — COMMENTED** [review]
Codewise it looks good to me :+1:

To your questions though:

-  I personally feel like a "Browser default" is a pretty decent default by itself, but if you would please share the business use case, probably it would help to understand what we are trying to solve here
- Default timezone should be probably go in hand with language, so if we decide to have one, we should have the other as well
- (do not know the answer) for Katello tests, I would solve it as the last thing, we can solve it even somewhere in `test_helper`.

File: app/models/user.rb [hunk 3]
```diff
@@ -25,6 +25,7 @@ class User < ApplicationRecord
   attribute :password
 
   after_save :ensure_default_role
+  after_initialize :set_default_language_for_new_user, if: :new_record?
```
Comment (ezr-ondrej) [inline app/models/user.rb]: By the way Rails 5 brought us:
```suggestion
  attribute :locale, :string, default: -> { Setting[:default_language].presence }
```
Haven't used it yet, do not know the pitfalls, but I think it is worth trying out, as it looks much cleaner :wink:

Comment (sbernhard) [reply app/models/user.rb]: will try. thanks for your suggestions.

**Comment by sbernhard** [comment]
Why is this a "nice" feature: 
A lot of clients complains, that they need to change the language to English for every user because foreman (including plugins like katello) aren't translated completely and the translations are often bad (= misleading; especially in case of technical terms). The clients browser is set to German, therefore German is used as the Browser locale. 

The story becomes even more important if you use LDAP / Active Directory, because here foreman-local users are generated "on the fly" using LDAP. 

There is another PR to set the locale / timezone for the initial admin user: #6736

**Review by ezr-ondrej — APPROVED** [review]
Makes sense, thanks for the explanation. I wonder if it would make sense to make the english default value than, but it would be more intrusive change.
This works nice and solves the issue, good job :+1:

We will just have to figure out what to do with the tests, the failing ones are not really written with flexibility in mind. But we can't break the test suite like that, so probably there will have to be a PR to the katello beforehand to prepare it. Would you mind to try doing that?

**Comment by sbernhard** [comment]
I can fix the tests on katello. Do you know if it's possible to link this PR to the fix katello tests PR? Otherwise it's maybe hard to run the tests.

**Comment by ezr-ondrej** [comment]
> I can fix the tests on katello. Do you know if it's possible to link this PR to the fix katello tests PR? Otherwise it's maybe hard to run the tests.

I do not know about that, but I would guess (sorry, only my guess here) we will need to wait till the PR on katello is merged, but if we are going to change just tests and link this PR, it should be fast.

**Comment by sbernhard** [comment]
[test foreman]

**Comment by ezr-ondrej** [comment]
[test katello]

**Comment by sbernhard** [comment]
@tbrisker can you please have a look on it again. Thanks.

**Review by tbrisker — APPROVED** [review]
I am indeed trying to reduce the number of settings, but I guess this one makes sense to add. Thanks @sbernhard!

**Comment by tbrisker** [comment]
Can you please also update the manual with these new settings, in https://github.com/theforeman/theforeman.org/blob/gh-pages/_includes/manuals/nightly/3.5.2_configuration_options.md#the-administersettings-page

**Pull request merged**